libbpf-tools/memleak.c

```diff
@@ -0,0 +1,707 @@
+			fprintf(stderr, "Invalid time: %s\n", arg);
+			argp_usage(state);
+		}
+		env.min_age_ns = older * 1e6;
+		break;
+	case 'C':
+		env.combined_only = true;
+		break;
+	case 'W':
+		env.wa_missing_free = true;
+		break;
+	case 's':
+		errno = 0;
+		sample_rate = strtol(arg, NULL, 10);
+		if (errno || sample_rate <= 0) {
+			fprintf(stderr, "Invalid sample rate: %s\n", arg);
+			argp_usage(state);
+		}
+		env.sample_rate = sample_rate;
+		break;
+	case 'z':
+		errno = 0;
+		min_size = strtol(arg, NULL, 10);
+		if (errno || min_size <= 0) {
+			fprintf(stderr, "Invalid min size: %s\n", arg);
+			argp_usage(state);
+		}
+		env.min_size = min_size;
+		break;
+	case 'Z':
+		errno = 0;
+		max_size = strtol(arg, NULL, 10);
+		if (errno || max_size <= 0) {
+			fprintf(stderr, "Invalid max size: %s\n", arg);
+			argp_usage(state);
+		}
+		env.max_size = max_size;
+		break;
+	case 'P':
+		env.percpu = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++ > 2) {
+			fprintf(stderr,
+				"Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		if (pos_args == 1) {
+			interval = strtoll(arg, NULL, 10);
+			if (errno || interval <= 0) {
+				fprintf(stderr, "Invalid report interval: %s\n", arg);
+				argp_usage(state);
+			}
+			env.interval = interval;
+		}
+		if (pos_args == 2) {
+			count = strtoll(arg, NULL, 10);
+			if (errno || count <= 0) {
+				fprintf(stderr, "Invalid number of times to print report: %s\n", arg);
+				argp_usage(state);
+			}
+			env.count = count;
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static const struct argp argp = {
+	.options = opts,
+	.parser = parse_arg,
+	.doc = argp_program_doc,
+};
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	return vfprintf(stderr, format, args);
+}
+
+static int get_libc_path(char *path)
+{
+	FILE *f;
+	char buf[PATH_MAX] = {};
+	char *filename;
+	float version;
+
+	f = fopen("/proc/self/maps", "r");
+	if (!f)
+		return -errno;
+
+	while (fscanf(f, "%*x-%*x %*s %*s %*s %*s %[^\n]\n", buf) != EOF) {
+		if (strchr(buf, '/') != buf)
+			continue;
+		filename = strrchr(buf, '/') + 1;
+		if (sscanf(filename, "libc-%f.so", &version) == 1) {
+			memcpy(path, buf, strlen(buf));
+			fclose(f);
+			return 0;
+		}
+	}
+
+	fclose(f);
+	return -1;
+}
+
+struct alloc_stack {
+	uint64_t stack_id;
+	int size;
+	int nr;
+};
+
+static int partition(struct alloc_stack *list, int low, int high)
+{
+	struct alloc_stack pivot = list[high];
+	struct alloc_stack temp;
+	int i = low - 1;
+	int j;
+
+	for (j = low; j <= high - 1; j++) {
+		if (list[j].size > pivot.size) {
+			i++;
+			temp = list[j];
+			list[j] = list[i];
+			list[i] = temp;
+		}
+	}
+
+	temp = list[i + 1];
+	list[i + 1] = list[high];
+	list[high] = temp;
+	return (i + 1);
+}
+
+static void quick_sort(struct alloc_stack *list, int low, int high)
```

Comment:
`qsort` in `stdlib.h` ?

Reply:
Fix in commit [2beb289e](https://github.com/iovisor/bcc/pull/3828/commits/2beb289e98d651bddc2127299c5627b5d6149fce).